### ISSUE TITLE
ErrorAnalysisDashboard issue

### ISSUE THREAD

**Issue by ashley8jain** [microsoft/responsible-ai-widgets#710]
```
from raiwidgets import ExplanationDashboard
ExplanationDashboard(global_explanation, model, dataset=X_test, true_y=Y_test) 
# no error occurred

from raiwidgets import ErrorAnalysisDashboard
ErrorAnalysisDashboard(global_explanation, model, dataset=X_test, true_y=Y_test, features=features)
# error occurred given below:
# ErrorAnalysisDashboard(global_explanation, model, dataset=X_test_list, true_y=y_test, features=features)
# ErrorAnalysisDashboard(global_explanation, model, dataset=np.array(X_test_list), true_y=y_test, features=features)
```

Error:-
```
/opt/conda/lib/python3.6/site-packages/pandas/core/base.py in item(self)
    658         if len(self) == 1:
    659             return next(iter(self))
--> 660         raise ValueError("can only convert an array of size 1 to a Python scalar")
    661 
    662     @property

ValueError: can only convert an array of size 1 to a Python scalar
```

**Comment by imatiach-msft** [comment]
hi @ashley8jain sorry about the trouble you are having.  Is there a longer stack trace you can provide?  I don't see what the stack is above this pandas method.  Also, is there some way to get a full repro notebook with some dummy dataset to debug this issue?  How is it that your model can accept a dataset as a list, as a numpy array, and as a pandas dataframe?  Does calling model.predict(X_test_list) actually work?

**Comment by ashley8jain** [comment]
@imatiach-msft 

`ErrorAnalysisDashboard(global_explanation, model, dataset=np.array(X_test_list), true_y=y_test, features=features)
`

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-33-4d42a7d73369> in <module>
      1 from raiwidgets import ErrorAnalysisDashboard
----> 2 ErrorAnalysisDashboard(global_explanation, model, dataset=np.array(X_test_list), true_y=Y_test)

~/.local/lib/python3.6/site-packages/raiwidgets/error_analysis_dashboard.py in __init__(self, explanation, model, dataset, true_y, classes, features, port, locale, public_ip, categorical_features, true_y_dataset, pred_y, model_task, metric, max_depth, num_leaves)
    153             public_ip=public_ip,
    154             port=port,
--> 155             locale=locale)
    156 
    157         if pred_y is None:

~/.local/lib/python3.6/site-packages/raiwidgets/dashboard.py in __init__(self, dashboard_type, model_data, public_ip, port, locale)
     74         self.add_route()
     75 
---> 76         html = self.load_index()
     77         print(f'{dashboard_type} started at {self._service.env.base_url}')
     78         self._service.env.display(html)

~/.local/lib/python3.6/site-packages/raiwidgets/dashboard.py in load_index(self)
     93 
     94     def load_index(self):
---> 95         index = self.load_widget_file("index.html")
     96         parser = InLineScript(self.load_widget_file)
     97         parser.feed(index)

~/.local/lib/python3.6/site-packages/raiwidgets/dashboard.py in load_widget_file(self, path)
    109                 "__rai_model_data__",
    110                 json.dumps(self.model_data,
--> 111                            default=serialize_json_safe))
    112             return content
    113 

/opt/conda/lib/python3.6/json/__init__.py in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    236         check_circular=check_circular, allow_nan=allow_nan, indent=indent,
    237         separators=separators, default=default, sort_keys=sort_keys,
--> 238         **kw).encode(obj)
    239 
    240 

/opt/conda/lib/python3.6/json/encoder.py in encode(self, o)
    197         # exceptions aren't as detailed.  The list call should be roughly
    198         # equivalent to the PySequence_Fast that ''.join() would do.
--> 199         chunks = self.iterencode(o, _one_shot=True)
    200         if not isinstance(chunks, (list, tuple)):
    201             chunks = list(chunks)

/opt/conda/lib/python3.6/json/encoder.py in iterencode(self, o, _one_shot)
    255                 self.key_separator, self.item_separator, self.sort_keys,
    256                 self.skipkeys, _one_shot)
--> 257         return _iterencode(o, 0)
    258 
    259 def _make_iterencode(markers, _default, _encoder, _indent, _floatstr,

~/.local/lib/python3.6/site-packages/responsibleai/serialization_utilities.py in serialize_json_safe(o)
     31         return serialize_json_safe(o.tolist())
     32     elif hasattr(o, 'item'):
---> 33         return o.item()  # numpy types
     34     elif hasattr(o, '__dict__'):
     35         return serialize_json_safe(o.__dict__)  # objects

/opt/conda/lib/python3.6/site-packages/pandas/core/base.py in item(self)
    658         if len(self) == 1:
    659             return next(iter(self))
--> 660         raise ValueError("can only convert an array of size 1 to a Python scalar")
    661 
    662     @property

ValueError: can only convert an array of size 1 to a Python scalar
```

yes, model.predict(np.array(X_test_list)) actually works.

**Comment by ashley8jain** [comment]
Alternate code:-

yes, model.predict(X_test) actually works.

```
model = xgboost.XGBClassifier(objective= 'binary:logistic', nthread=4, early_stopping_rounds=10, scale_pos_weight = 1, max_depth=1, min_child_weight=7, gamma=0.0, subsample=0.7, colsample_bytree=1.0, reg_alpha=1, learning_rate=0.05, n_estimators=500)

model.fit(X_train, Y_train) # X_train is panda dataframe with column headers and Y_train is series
model.predict(X_test)
proba_target_list_test = model.predict_proba(X_test)
explainer = MimicExplainer(model, 
                           X_train, 
                           LinearExplainableModel)
global_explanation = explainer.explain_global(X_test)
# Sorted SHAP values
print('ranked global importance values: {}'.format(global_explanation.get_ranked_global_values()))
# Corresponding feature names
print('ranked global importance names: {}'.format(global_explanation.get_ranked_global_names()))
# Feature ranks (based on original order of features)
print('global importance rank: {}'.format(global_explanation.global_importance_rank))

# Note: Do not run this cell if using PFIExplainer, it does not support per class explanations
# Per class feature names
print('ranked per class feature names: {}'.format(global_explanation.get_ranked_per_class_names()))
# Per class feature importance values
print('ranked per class feature values: {}'.format(global_explanation.get_ranked_per_class_values()))
print('global importance rank: {}'.format(global_explanation.get_feature_importance_dict()))

ErrorAnalysisDashboard(global_explanation, model, dataset=X_test, true_y=Y_test) # error occurred at this line
```

Error (occurred at last line of above code):-
```
ValueError                                Traceback (most recent call last)
<ipython-input-33-4d42a7d73369> in <module>
      1 from raiwidgets import ErrorAnalysisDashboard
----> 2 ErrorAnalysisDashboard(global_explanation, model, dataset=np.array(X_test_list), true_y=Y_test)

~/.local/lib/python3.6/site-packages/raiwidgets/error_analysis_dashboard.py in __init__(self, explanation, model, dataset, true_y, classes, features, port, locale, public_ip, categorical_features, true_y_dataset, pred_y, model_task, metric, max_depth, num_leaves)
    153             public_ip=public_ip,
    154             port=port,
--> 155             locale=locale)
    156 
    157         if pred_y is None:

~/.local/lib/python3.6/site-packages/raiwidgets/dashboard.py in __init__(self, dashboard_type, model_data, public_ip, port, locale)
     74         self.add_route()
     75 
---> 76         html = self.load_index()
     77         print(f'{dashboard_type} started at {self._service.env.base_url}')
     78         self._service.env.display(html)

~/.local/lib/python3.6/site-packages/raiwidgets/dashboard.py in load_index(self)
     93 
     94     def load_index(self):
---> 95         index = self.load_widget_file("index.html")
     96         parser = InLineScript(self.load_widget_file)
     97         parser.feed(index)

~/.local/lib/python3.6/site-packages/raiwidgets/dashboard.py in load_widget_file(self, path)
    109                 "__rai_model_data__",
    110                 json.dumps(self.model_data,
--> 111                            default=serialize_json_safe))
    112             return content
    113 

/opt/conda/lib/python3.6/json/__init__.py in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    236         check_circular=check_circular, allow_nan=allow_nan, indent=indent,
    237         separators=separators, default=default, sort_keys=sort_keys,
--> 238         **kw).encode(obj)
    239 
    240 

/opt/conda/lib/python3.6/json/encoder.py in encode(self, o)
    197         # exceptions aren't as detailed.  The list call should be roughly
    198         # equivalent to the PySequence_Fast that ''.join() would do.
--> 199         chunks = self.iterencode(o, _one_shot=True)
    200         if not isinstance(chunks, (list, tuple)):
    201             chunks = list(chunks)

/opt/conda/lib/python3.6/json/encoder.py in iterencode(self, o, _one_shot)
    255                 self.key_separator, self.item_separator, self.sort_keys,
    256                 self.skipkeys, _one_shot)
--> 257         return _iterencode(o, 0)
    258 
    259 def _make_iterencode(markers, _default, _encoder, _indent, _floatstr,

~/.local/lib/python3.6/site-packages/responsibleai/serialization_utilities.py in serialize_json_safe(o)
     31         return serialize_json_safe(o.tolist())
     32     elif hasattr(o, 'item'):
---> 33         return o.item()  # numpy types
     34     elif hasattr(o, '__dict__'):
     35         return serialize_json_safe(o.__dict__)  # objects

/opt/conda/lib/python3.6/site-packages/pandas/core/base.py in item(self)
    658         if len(self) == 1:
    659             return next(iter(self))
--> 660         raise ValueError("can only convert an array of size 1 to a Python scalar")
    661 
    662     @property

ValueError: can only convert an array of size 1 to a Python scalar
```

**Comment by MasaNahti** [comment]
Hi,

I faced exactly the same issue a few days ago. My `X_test` was a pandas dataframe and `y_test` pandas series. Calling `tolist()` on `y_test` as shown below solved the issue:

`ErrorAnalysisDashboard(global_explanation, model, dataset=X_test, true_y=y_test.tolist())
`

I gave it a try after reading [this answer](https://stackoverflow.com/questions/61670898/pandas-can-only-convert-an-array-of-size-1-to-a-python-scalar#answer-61673971) on Stackoverflow about this particular ValueError.

**Comment by imatiach-msft** [comment]
@MasaNahti ah I see, great catch.  Yes pandas series is not supported as an input, and is not tested anywhere.  Maybe I should add support for this type of array.

**Comment by MasaNahti** [comment]
Thank you for your quick reply! Well, it _is_ written in the docs that the `true_y` is either numpy array or list, so this isn't really a bug. It is just that after getting used to pandas dataframes and series with sklearn and interpretML, it can be slightly confusing. 

But I just remembered what confused me most. Before diving into Error Analysis, I followed [this instruction](https://github.com/interpretml/interpret-community/blob/master/notebooks/explain-regression-local.ipynb) from interpret-community repo to use their TabularExplainer. In the end, the results are visualized with an ExplanationDashboard from rai-widgets. This is the interesting part now: I used _the same_ `y_test` - a pandas series - that caused issues in ErrorAnalysisDashboard before transformed to list, but here there was no error! The dashboard worked without any issue, even though the docs state that `true_y` should be a numpy array or a list as well. What am I missing here?

So, to sum up, this works:
`ExplanationDashboard(global_explanation, model, dataset=X_test, true_y=y_test)`

And this doesn't: 
`ErrorAnalysisDashboard(global_explanation, model, dataset=X_test, true_y=y_test)`

For the same `y_test` - pandas Series.

**Comment by imatiach-msft** [comment]
@MasaNahti yes it seems even though it's not supported officially and not tested it is handled in the ExplanationDashboard here:

https://github.com/microsoft/responsible-ai-widgets/blob/main/raiwidgets/raiwidgets/explanation_dashboard_input.py#L285

but ErrorAnalysisDashboard defines the same method but has this line missing:

https://github.com/microsoft/responsible-ai-widgets/blob/main/raiwidgets/raiwidgets/error_analysis_dashboard_input.py#L513

I think (1) this method should be refactored out to common place so all dashboard use them (2) we should add tests for pandas series type everywhere and add to our docs that we support it for the prediction/label columns

**Comment by MasaNahti** [comment]
@imatiach-msft great observation! I believe your suggestions would indeed make it all a bit more transparent. 

@ashley8jain could you check if my solution works for you as well? (It should, really.) Afterwards this particular issue could be closed. Also the "bug" label should perhaps be removed.

**Comment by imatiach-msft** [comment]
closing as fixed in PR #815
